### PR TITLE
[GSoC] Migrated RepairCollection to Coroutines

### PR DESCRIPTION
**Pull request by divyansh-dxn**
## Fixes
Fixes a part of #7108

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration (SDK version(s), emulator or physical, etc)

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt [hunk 1]
```diff
@@ -491,16 +491,6 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
         }
     }
 
-    @KotlinCleanup("doesn't work on null collection - only on non-openable")
-    class RepairCollection : UnsafeTaskDelegate<Void, Boolean>() {
-        override fun task(col: Collection, collectionTask: ProgressSenderAndCancelListener<Void>): Boolean {
-            Timber.d("doInBackgroundRepairCollection")
-            Timber.i("RepairCollection: Closing collection")
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:498]: Also add in your code the two Timber calls.

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt [hunk 2]
```diff
@@ -491,16 +491,6 @@ open class CollectionTask<Progress, Result>(val task: TaskDelegateBase<Progress,
         }
     }
 
-    @KotlinCleanup("doesn't work on null collection - only on non-openable")
-    class RepairCollection : UnsafeTaskDelegate<Void, Boolean>() {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:495]: `UnsafeTaskDelegate` becomes unused as `RepairCollection` was its only child so delete that class as well.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 3]
```diff
@@ -1425,33 +1425,20 @@ open class DeckPicker :
...
-        }
-    }
-
     // Callback method to handle repairing deck
     fun repairCollection() {
         Timber.i("Repairing the Collection")
-        TaskManager.launchCollectionTask(RepairCollection(), repairCollectionTask())
+        launchCatchingTask(resources.getString(R.string.deck_repair_error)) {
+            val result = withProgress(resources.getString(R.string.backup_repair_deck_progress)) {
+                withCol {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1433]: Add the cleanup from `RepairCollection` as a todo here:

_// todo doesn't work on null collection-only on non-openable(still applicable with withCol?)_

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 4]
```diff
@@ -1425,33 +1425,20 @@ open class DeckPicker :
...
-            )
-        }
-
-        override fun actualOnPostExecute(context: DeckPicker, result: Boolean?) {
-            if (context.mProgressDialog != null && context.mProgressDialog!!.isShowing) {
-                context.mProgressDialog!!.dismiss()
-            }
-            if (!result!!) {
-                showThemedToast(context, context.resources.getString(R.string.deck_repair_error), true)
-                context.showCollectionErrorDialog()
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1446]: This method call is not present in your code when result is false.

**Review by lukstbit — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 4 above)
Comment (divyansh-dxn) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1446]: https://github.com/ankidroid/Anki-Android/pull/12498/files/15d4976693082a8fbec3b93693ccdc259b81e796#diff-1c9f8440ee3813eda89e500bf5db95d90730eda462a22fb54b2d832e8fb06e2cR1438-R1440

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 3 above)
Comment (divyansh-dxn) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1435]: done

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt (hunk 2 above)
Comment (divyansh-dxn) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:495]: done

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 4 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1446]: I don't understand your answer @divyansh-dxn . The line you selected in your link contains the showThemedToast but not the `context.showCollectionErrorDialog()`

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:498]: Please wait until you have uploaded the code with the correction to resolve. It saves me time, as I was going to note the same things.
Or at least to ask you why you deleted those Timber; it may be justified, but have to be explained

**Review by Arthur-Milchior — DISMISSED** [review]
Still one question.
I guess we should squash

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 4 above)
Comment (divyansh-dxn) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1446]: Sorry I missed that one, I thought of showThemedToast only but indeed showCollectionErrorDialog is an important part of the failure because it shows what can be done to fix this.

**Review by lukstbit — APPROVED** [review]

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt (hunk 1 above)
Comment (divyansh-dxn) [reply AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.kt:498]: Squashed the changes ...will be on the same commit now, so no issues.

**Review by ShridharGoel — APPROVED** [review]
